src/pwm.rs

```diff
@@ -153,54 +154,130 @@ pub trait Pins<TIM> {
+  the same thing
+
+  ```rust
+    fn impl_pins(pin_channel_mapping: &[Option<(Channel, Pin), Unused> {}
+
+    fn impl_pin_combinations(pins: &[Pin]) {
+        fn internal(
+            expanded: &[Option<(Channel, Pin)pins, Unused>],
+            unexpanded: [(Pin, Channel)]
+        ) -> impls {
+            if unexpanded.is_empty() {
+                impl_pins(expanded)
+            }
+            else {
+                let (first, rest) = split(unexpanded);
+                // Use the pin
+                internal(expanded ++ first, unexpanded);
+                // Don't use the pin
+                internal(expanded ++ Unused{}, unexpanded);
+            }
+        }
+        let pins_with_channels = pins.zip([C1, C2, C3, C4]);
+    }
+  ```
+*/
+macro_rules! impl_pin_combinations {
+    // Base case, all pins have been expanded
+    (
+        $( $channel:ident($($pin:ident)? ; $($unused_type:ty)?), )* ;
+        ; for $timX:ident, $remap:expr
+    ) => {
+        impl_pins!{
+            $( $channel( $($channel, $pin, true)?; $($unused_type, false)? ) ),*
+            ; for $timX, $remap
+        }
+    };
+    // Intermediate recursion. Expand the first non-expanded pin/channel combo
+    // to be enabled and disabled
+    (
+        $( $channel:ident($($pin:ident)? ; $($unused_type:ty)?), )*
+        ;
+        ($first_channel:ident,$first_pin:ident),$( ($rest_channel:ident, $rest_pin:ident), )*
+        ; for $timX:ident, $remap:expr
+    ) => {
+        // Expand with pin enabled
+        impl_pin_combinations! {
+            // Add the previously expanded pins
+            $($channel($($pin)? ; $($unused_type)? ),)*
+            $first_channel($first_pin;),
+            ;
+            $( ($rest_channel, $rest_pin), )*
+            ; for $timX, $remap
+        }
+        // Expand with pin disabled
+        impl_pin_combinations! {
+            // Add the previously expanded pins
+            $($channel($($pin)? ; $($unused_type)? ),)*
+            $first_channel(;()),
+                ;
+            $( ($rest_channel, $rest_pin), )*
+            ; for $timX, $remap
+        }
+    };
+    // No expanded pins, call recursive version
+    ( $p1:ident, $p2:ident, $p3:ident, $p4:ident ; for $timX:ident, remap $remap:expr) => {
+        impl_pin_combinations!{
+            ; (C1, $p1), (C2, $p2), (C3, $p3), (C4, $p4),
+            ; for $timX, $remap
+        }
+    }
 }
 
+
+impl_pin_combinations!(PA0,  PA1, PA2,  PA3; for TIM2,  remap 0b00);
+impl_pin_combinations!(PA15, PB3, PA2,  PA3; for TIM2,  remap 0b01);
+impl_pin_combinations!(PA0,  PA1, PB10, PB11; for TIM2, remap 0b01);
+impl_pin_combinations!(PA15, PB3, PB10, PB11; for TIM2, remap 0b01);
+
+impl_pin_combinations!(PA6, PA7, PB0, PB1; for TIM3, remap 0b00);
+impl_pin_combinations!(PB4, PB5, PB0, PB1; for TIM3, remap 0b01);
+impl_pin_combinations!(PC6, PC7, PC8, PC9; for TIM3, remap 0b11);
+
+impl_pin_combinations!(PB6,  PB7,  PB8,  PB9 ; for TIM4, remap 0b0);
```

Comment:
Is there any reason you want to keep 'for' here ?

Reply:
Not really, I just feelt like it makes it slightly more readable